### PR TITLE
🌱 Update contract package to match documentation

### LINKED ISSUE #5309: Extend usage of the contract package
During the work for ClusterClass we started encoding the Cluster API contract as a set of golang types wrapping unstructured defined in https://github.com/kubernetes-sigs/cluster-api/tree/main/controllers/topology/internal/contract.

This not only provided a cleaner code in the topology controller, but it also provides a well defined, centralized representation of the Cluster API contract.

Given those positive results we should now consider if to refactor the rest of the codebase to use the contract package, thus removing all the contract related operations with unstructured spread across the codebase.

This could require to:
- Move the contract package to a different location (depending on where the cleanup will impact to)
- Extending the contract package with all the rules not yet implemented (all the rules not relevant for topology reconciliation)
- Identify and modify all the places in the codebase where we rely on the Cluster API contract assumptions (a good starting point are the "reconcincileExternal" methods in the controller, but there could be other points as well)

Note: this work could eventually be divided into a series of PRs.

/kind feature

### LINKED ISSUE #7119: Implement the bootstrap provider contract in the contract package
Implement the bootstrap provider contract outlined [here](https://cluster-api.sigs.k8s.io/developer/providers/bootstrap.html) as a sub issue in #5309.

### PR DESCRIPTION
**Pull request by Jont828**
**What this PR does / why we need it**: As described in #5309, we're going through `internal/contract` package and updating it to match the specifications in the docs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7119 #7120

### CODE REVIEW

**Comment by Jont828** [comment]
@killianmuldoon @fabriziopandini

**Review by killianmuldoon — COMMENTED** [review]
Looks like a good start - could you add some unit tests to show these fields working?

**Comment by sbueringer** [comment]
/ok-to-test

I would take a look once the tests are green / linter is happy

**Comment by Jont828** [comment]
I held off on tests since I had a couple design considerations I wanted to ask about.

1. I noticed we're created an Int64` type to represent in `Int32` fields, was that an intentional change to group them all as `Int64`?
2. I think it would be helpful to documents optional fields as well, but currently there isn't a good way to show the difference between an optional field like `failureMessage` and a required field like `ready`. 
3. At what point do we want to create additional structs for nested fields? For example, `InfrastructureClusterContract` returns a field `InfrastructureClusterControlPlaneEndpoint` which has additional fields so that it would be called as `infra.ControlPlaneEndpoint().Host()`. Do we want every nested path to have a intermediate struct like `infra.Status().Ready()` or just have it be `infra.Ready()`?

**Comment by sbueringer** [comment]
> @sbueringer @killianmuldoon I held off on tests since I had a couple design considerations I wanted to ask about.
> 
> 1. I noticed we're created an `Int64` type to represent in `Int32` fields, was that an intentional change to group them all as `Int64`?
> 2. I think it would be helpful to documents optional fields as well, but currently there isn't a good way to show the difference between an optional field like `failureMessage` and a required field like `ready`.
> 3. At what point do we want to create additional structs for nested fields? For example, `InfrastructureClusterContract` returns a field `InfrastructureClusterControlPlaneEndpoint` which has additional fields so that it would be called as `infra.ControlPlaneEndpoint().Host()`. Do we want every nested path to have a intermediate struct like `infra.Status().Ready()` or just have it be `infra.Ready()`?

* 1. I think this is due to:
  * The contract (`control-plane.md`) just specifices that e.g. `replicas`  is an int. Not if it's int32 or int64. I think int64 is just safer
  * The unstructured package only has a func for int64: `unstructured.NestedInt64`
* 2. I think it's fine to have the same funcs as in the usual cases and if the field is not set we just return an error that wraps `errNotFound` like we do today in other funcs (e.g. .status.replicas is also not always set)
* 3. In general we want the full path and everything except the leaf should be a struct

**Comment by Jont828** [comment]
Gotcha that makes sense to me. I also spoke with @jackfrancis and we were also wondering if there's a way to generate code for this as opposed to manually implementing the paths and structs.

File: internal/contract/infrastructure_cluster.go [hunk 1]
```diff
@@ -61,6 +61,79 @@ func (c *InfrastructureClusterControlPlaneEndpoint) Port() *Int64 {
...
+	}
+}
+
+// Option 1 for FailureDomains:
+// Pros: Provides granularity for accessing map[string]FailureDomainSpec using intermediate structs, all `path` return
+// types are wrappers for primatives instead of custom structs.
+// Cons: Accessing the map[string]FailureDomainSpec requires knowing the keys ahead of time.
+
+// FailureDomains provides access to the status.failureDomains field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureDomains() *InfrastructureClusterFailureDomains {
```
Comment (Jont828) [inline internal/contract/infrastructure_cluster.go:91]: Here's one way we could implement `FailureDomains()` which is a map[string]FailureDomainSpec as outlined here. I implemented two options in the PR, in this one we'd need to know the key name ahead of time to use `FailureDomainSpec`.

File: internal/contract/infrastructure_cluster.go [hunk 2]
```diff
@@ -61,6 +61,79 @@ func (c *InfrastructureClusterControlPlaneEndpoint) Port() *Int64 {
...
+		path: []string{"status", "failureDomains", c.Name, "attributes"},
+	}
+}
+
+// Option 2 for FailureDomains:
+// Pros: Directly returns the map[string]FailureDomainSpec, allowing users to determine the keys at runtime.
+// Cons: Requires defining a new intermediate struct and doesn't document the subfields within a FailureDomainSpec.
+
+// FailureDomains provides access to the status.failureDomains field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureDomainsAlt() *FailureDomains {
```
Comment (Jont828) [inline internal/contract/infrastructure_cluster.go:131]: Here's the second way we could implement it such that we just return the `map[string]FailureDomainSpec` directly. While this method doesn't require knowing the keys ahead of time, it does require defining a new `FailureDomains` type with the getter and setter.

File: internal/contract/infrastructure_machine.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+// FailureMessage provides access to the status.failureMessage field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureMessage() *String {
+	return &String{
+		path: []string{"status", "failureMessage"},
+	}
+}
+
+// Addresses provides access to the status.addresses field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) Addresses() *MachineAddresses {
```
Comment (Jont828) [inline internal/contract/infrastructure_machine.go:60]: The [contract](https://cluster-api.sigs.k8s.io/developer/providers/machine-infrastructure.html) specifies that `status.addresses` field is a list of MachineAddress structs. Is there a way to access indices in a list using the JSON path syntax we've outlined here? Currently I've just made status.addresses a leaf node and that returns a new MachineAddress type I've defined. I'm definitely open to suggestions on how to handle lists in the contract. While there is an unstructured.NestedSlice method, I had to define a new type to implement the Set() and Get() methods.

**Review by Jont828 — COMMENTED** [review]
@sbueringer I've worked on implementing the optional types but had some trouble dealing with types that were a list or a map of additional structs. Would definitely appreciate any suggestions on how to proceed!

**Comment by sbueringer** [comment]
I'll take a look early next week

**Comment by sbueringer** [comment]
> Gotcha that makes sense to me. I also spoke with @jackfrancis and we were also wondering if there's a way to generate code for this as opposed to manually implementing the paths and structs.

Somehow forgot to answer. I think it might be possible but I wonder if maintaining the generation code (+ unit tests, ..) is worth it.

File: internal/contract/infrastructure_cluster.go (hunk 2 above)
Comment (sbueringer) [reply internal/contract/infrastructure_cluster.go:131]: I would prefer this option. I think both options require to define a type (above it was `InfrastructureClusterFailureDomains` here it is `FailureDomains`).

Feels more straightforward to me to just return the entire clusterv1.FailureDomains structure

Comment (sbueringer) [reply internal/contract/infrastructure_cluster.go:131]: @fabriziopandini WDYT?

File: internal/contract/infrastructure_machine.go (hunk 3 above)
Comment (sbueringer) [reply internal/contract/infrastructure_machine.go]: Same as in https://github.com/kubernetes-sigs/cluster-api/pull/7138#discussion_r976269146

I would just let Addresses() get/set the entire slice. It's a lot easier to handle then reimplemting slice insert/remove/update in the contract package

File: internal/contract/infrastructure_machine.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// ProviderID provides access to the spec.providerID field in an InfrastructureMachine object.
+func (m *InfrastructureMachineContract) ProviderID() *String {
+	return &String{
+		path: []string{"spec", "providerID"},
+	}
+
+}
+
+// FailureDomain provides access to the spec.failureDomain field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureDomain() *String {
```
Comment (fabriziopandini) [inline internal/contract/infrastructure_machine.go:108]: This is not documented in https://cluster-api.sigs.k8s.io/developer/architecture/controllers/machine.html; could you. kindly open an issue to fix this as well as documenting Hostname and Addresses

File: internal/contract/types.go [hunk 5]
```diff
@@ -166,3 +198,124 @@ func (i *Duration) Set(obj *unstructured.Unstructured, value metav1.Duration) er
 	}
 	return nil
 }
+
+// MachineAddresses represents an accessor to a []clusterv1.MachineAddress path value.
+type MachineAddresses struct {
```
Comment (fabriziopandini) [inline internal/contract/types.go:203]: If this type is specific only of infra machine, let's consider to move the implementation in the corresponding file (same for FailureDomains below)

**Review by fabriziopandini — COMMENTED** [review]
Made a first pass, this is a great work of research.
I'm looking forward to get this merged and start cleaning up usage of unstructured all around the codebase

File: internal/contract/infrastructure_cluster.go [hunk 6]
```diff
@@ -61,6 +61,79 @@ func (c *InfrastructureClusterControlPlaneEndpoint) Port() *Int64 {
...
+}
+
+// FailureMessage provides access to the status.failureMessage field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureMessage() *String {
+	return &String{
+		path: []string{"status", "failureMessage"},
+	}
+}
+
+// Option 1 for FailureDomains:
```
[the remark this answers is not in the document]
Comment (Jont828) [reply internal/contract/infrastructure_cluster.go:85]: Sure thing!

File: internal/contract/infrastructure_cluster.go [hunk 7]
```diff
@@ -61,6 +61,79 @@ func (c *InfrastructureClusterControlPlaneEndpoint) Port() *Int64 {
...
+
+// Attributes provides access to the status.failureDomains.<name>.attributes field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterFailureDomainSpec) Attributes() *StringMap {
+	return &StringMap{
+		path: []string{"status", "failureDomains", c.Name, "attributes"},
+	}
+}
+
+// Option 2 for FailureDomains:
+// Pros: Directly returns the map[string]FailureDomainSpec, allowing users to determine the keys at runtime.
```
[the remark this answers is not in the document]
Comment (Jont828) [reply internal/contract/infrastructure_cluster.go:127]: Sounds good, I'll refactor this PR to use option 2 for this and `MachineAddresses`

File: internal/contract/infrastructure_machine.go (hunk 4 above)
Comment (sbueringer) [reply internal/contract/infrastructure_machine.go:108]: Done

File: internal/contract/infrastructure_machine.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		return nil, errors.Wrapf(err, "failed to unmarshall field at %s to json", "."+strings.Join(m.path, "."))
+	}
+
+	return &addresses, nil
+}
+
+// Set sets the []clusterv1.MachineAddress value in the path.
+func (m *MachineAddresses) Set(obj *unstructured.Unstructured, values []clusterv1.MachineAddress) error {
+	slice := make([]interface{}, len(values))
+	s, err := json.Marshal(values)
```
Comment (Jont828) [inline internal/contract/infrastructure_machine.go:123]: @sbueringer @fabriziopandini One thing I am curious about is if there's a better way to implement the Set() and Get() methods for nested structs. Right now I'm just marshalling the value into a string and unmarshalling it into either the struct or a interface.

Comment (fabriziopandini) [reply internal/contract/infrastructure_machine.go:123]: I don't think so. Basically, we are mimicking what unstructured does; AFIAK also other implementations like duck typing in KNative use a similar approach.

Comment (sbueringer) [reply internal/contract/infrastructure_machine.go:123]: Doesn't look like it.

Wondering if we should use `SetNestedSlice` though

Comment (Jont828) [reply internal/contract/infrastructure_machine.go:123]: Gotcha I'll just leave it as is then. Also [SetNestedSlice()](https://github.com/kubernetes/apimachinery/blob/v0.25.2/pkg/apis/meta/v1/unstructured/helpers.go#L244) is implemented just as a wrapper for SetNestedField() so I figured it's all the same.

**Comment by Jont828** [comment]
@fabriziopandini @sbueringer I squashed and fixed the linting issues, PTAL when you get the chance!

**Comment by Jont828** [comment]
@fabriziopandini @sbueringer I think this should be ready for a final pass!

**Comment by sbueringer** [comment]
Thx!

I'll take a look as soon as I can, but I can't promise it will happen before KubeCon. Just too much things going on right now.

**Review by fabriziopandini — COMMENTED** [review]
I'm +1 to merge this code, so we can start using it as a replacement for current unstructured handling across the code base.
/lgtm

**Comment by Jont828** [comment]
@fabriziopandini @sbueringer Are we good to merge this or do we need a final pass of reviews?

**Comment by sbueringer** [comment]
Looks good to me!

/approve

**Pull request merged**